File: AID/r/aid.R
# ============================================================================
# AID (Automatic Interaction Detection) Regressor in R — SSE / least squares
# ============================================================================
# Key fixes vs your previous version:
# - Correct split routing: left/right decided with split_value (midpoint), not with X_sorted[best_idx]
# - Faster optimal split scan using cumulative sums (O(n) after sorting)
# - Stop condition includes feasibility of two leaves of size >= min_samples_leaf
# - Optional deterministic feature subsampling (max_features) via set.seed(random_state) at fit()
# ============================================================================

AIDNode <- setRefClass("AIDNode",
  fields = list(
    prediction = "numeric",
    sse = "numeric",
    n_samples = "numeric",
    split_feature_idx = "numeric",
    split_value = "numeric",
    split_gain = "numeric",
    f_statistic = "numeric",
    left = "ANY",
    right = "ANY",
    is_leaf = "logical",
    depth = "numeric"
  ),
  methods = list(
    initialize = function() {
      prediction <<- 0.0
      sse <<- 0.0
      n_samples <<- 0
      split_feature_idx <<- NA_integer_
      split_value <<- NA_real_
      split_gain <<- 0.0
      f_statistic <<- NA_real_
      left <<- NULL
      right <<- NULL
      is_leaf <<- TRUE
      depth <<- 0
    }
  )
)

AIDRegressor <- setRefClass("AIDRegressor",
  fields = list(
    min_samples_split = "numeric",
    min_samples_leaf = "numeric",
    max_depth = "numeric",
    min_split_gain = "numeric",
    max_features = "numeric",
    store_history = "logical",
    max_leaves = "numeric",
    random_state = "numeric",

    root_ = "ANY",
    n_features_ = "numeric",
    feature_names_ = "character",
    X_ = "matrix",
    y_ = "numeric",
    history_ = "list",
    n_leaves_ = "numeric"
  ),

  methods = list(
    initialize = function(min_samples_split = 10,
                          min_samples_leaf = 5,
                          max_depth = 5,
                          min_split_gain = 0.0,
                          max_features = NA,
                          store_history = FALSE,
                          max_leaves = Inf,
                          random_state = 0) {
      min_samples_split <<- as.numeric(min_samples_split)
      min_samples_leaf <<- as.numeric(min_samples_leaf)
      max_depth <<- as.numeric(max_depth)
      min_split_gain <<- as.numeric(min_split_gain)
      max_features <<- max_features
      store_history <<- store_history
      max_leaves <<- as.numeric(max_leaves)
      random_state <<- as.numeric(random_state)

      root_ <<- NULL
      n_features_ <<- 0
      feature_names_ <<- character(0)
      X_ <<- matrix(numeric(0), nrow = 0, ncol = 0)
      y_ <<- numeric(0)
      history_ <<- list()
      n_leaves_ <<- 0
    },

    validate_input = function(X, y) {
      if (is.data.frame(X)) {
        feature_names_ <<- colnames(X)
        X <- as.matrix(X)
      } else {
        X <- as.matrix(X)
        feature_names_ <<- paste0("X", seq_len(ncol(X)))
      }

      storage.mode(X) <- "double"
      y <- as.numeric(y)

      if (nrow(X) != length(y)) {
        stop(sprintf("Shape mismatch: X has %d samples, y has %d", nrow(X), length(y)))
      }

      if (any(is.na(X)) || any(is.infinite(X))) stop("X contains NA or Inf")
      if (any(is.na(y)) || any(is.infinite(y))) stop("y contains NA or Inf")

      list(X = X, y = y)
    },

    fit = function(X, y) {
      result <- validate_input(X, y)
      X <- result$X
      y <- result$y

      X_ <<- X
      y_ <<- y
      n_features_ <<- ncol(X_)

      if (is.na(max_features)) {
        max_features <<- n_features_
      } else {
        max_features <<- min(as.numeric(max_features), n_features_)
      }

      set.seed(as.integer(random_state))

      sorted_indices_ <<- lapply(seq_len(n_features_), function(j) {
        order(X_[, j], method = "radix")
      })

      history_ <<- list()
      n_leaves_ <<- 0

      indices <- seq_len(length(y_))
      root_ <<- build_tree(indices, depth = 0)

      invisible(.self)
    },

    find_numeric_split = function(indices, feature_idx, parent_sse) {
    
      global_ord <- sorted_indices_[[feature_idx]]
      
      in_node <- logical(nrow(X_))
      in_node[indices] <- TRUE
      
      idx_sorted_global <- global_ord[in_node[global_ord]]
      
      
      x_sorted <- X_[idx_sorted_global, feature_idx]
      y_sorted <- y_[idx_sorted_global]
      
      n <- length(y_sorted)
      if (n < 2 * min_samples_leaf) return(NULL)

      diff_mask <- x_sorted[-n] != x_sorted[-1]
      
      if (!any(diff_mask)) return(NULL)
      split_pos <- which(diff_mask)  

      left_n <- split_pos
      right_n <- n - split_pos
      valid <- (left_n >= min_samples_leaf) & (right_n >= min_samples_leaf)
      split_pos <- split_pos[valid]
      if (length(split_pos) == 0) return(NULL)

      # cumulative sums for fast SSE
      csum <- cumsum(y_sorted)
      csum2 <- cumsum(y_sorted * y_sorted)
      total_sum <- csum[n]
      total_sum2 <- csum2[n]

      left_sum <- csum[split_pos]
      left_sum2 <- csum2[split_pos]
      right_sum <- total_sum - left_sum
      right_sum2 <- total_sum2 - left_sum2

      left_n_f <- as.numeric(split_pos)
      right_n_f <- as.numeric(n - split_pos)

      left_sse <- left_sum2 - (left_sum * left_sum) / left_n_f
      right_sse <- right_sum2 - (right_sum * right_sum) / right_n_f

      within_sse <- left_sse + right_sse
      gains <- parent_sse - within_sse

      best_i <- which.max(gains)
      best_gain <- gains[best_i]
      if (is.na(best_gain) || best_gain <= 0) return(NULL)

      k <- split_pos[best_i]
      split_value <- (x_sorted[k] + x_sorted[k + 1]) / 2.0

      denom <- within_sse[best_i] / max(n - 2, 1)
      f_stat <- if (denom > 0) best_gain / denom else Inf

      left_idx <- idx_sorted_global[1:k]
      right_idx <- idx_sorted_global[(k + 1):n]

      list(
        split_value = as.numeric(split_value),
        gain = as.numeric(best_gain),
        f_stat = as.numeric(f_stat),
        left_idx = left_idx,
        right_idx = right_idx
      )
    },

    should_stop = function(node, depth) {
      if (depth >= max_depth) return(TRUE)
      if (node$n_samples < min_samples_split) return(TRUE)
      if (n_leaves_ >= max_leaves) return(TRUE)
      if (node$n_samples < 2 * min_samples_leaf) return(TRUE)
      FALSE
    },

    build_tree = function(indices, depth) {
      node <- AIDNode$new()
      node$depth <- depth
      node$n_samples <- length(indices)

      y_subset <- y_[indices]
      node$prediction <- mean(y_subset)
      # SSE using centered sum-of-squares
      node$sse <- sum((y_subset - node$prediction)^2)

      if (should_stop(node, depth)) {
        n_leaves_ <<- n_leaves_ + 1
        return(node)
      }

      best_split <- find_best_split(indices, node$sse)

      if (is.null(best_split) || best_split$gain <= min_split_gain) {
        n_leaves_ <<- n_leaves_ + 1
        return(node)
      }

      left_idx <- best_split$left_idx
      right_idx <- best_split$right_idx

      if (length(left_idx) < min_samples_leaf || length(right_idx) < min_samples_leaf) {
        n_leaves_ <<- n_leaves_ + 1
        return(node)
      }

      node$is_leaf <- FALSE
      node$split_feature_idx <- best_split$feature_idx
      node$split_value <- best_split$split_value
      node$split_gain <- best_split$gain
      node$f_statistic <- best_split$f_stat

      if (store_history) {
        history_[[length(history_) + 1]] <<- list(
          depth = depth,
          feature_idx = best_split$feature_idx,
          split_value = best_split$split_value,
          gain = best_split$gain,
          f_statistic = best_split$f_stat
        )
      }

      node$left <- build_tree(left_idx, depth + 1)
      node$right <- build_tree(right_idx, depth + 1)
      node
    },

    find_best_split = function(indices, parent_sse) {
      if (length(indices) < 2 * min_samples_leaf) return(NULL)

      # choose subset of features (optional)
      features_to_try <- if (max_features < n_features_) {
        sample(seq_len(n_features_), size = as.integer(max_features), replace = FALSE)
      } else {
        seq_len(n_features_)
      }

      best_gain <- -Inf
      best_result <- NULL

      for (feature_idx in features_to_try) {
        result <- find_numeric_split(indices, feature_idx, parent_sse)
        if (!is.null(result) && result$gain > best_gain) {
          best_gain <- result$gain
          best_result <- c(list(feature_idx = feature_idx), result)
        }
      }

      best_result
    },

    find_numeric_split = function(indices, feature_idx, parent_sse) {
      # sort indices by feature value
      x <- X_[indices, feature_idx]
      ord <- order(x, method = "radix")
      x_sorted <- x[ord]
      y_sorted <- y_[indices][ord]
      idx_sorted_global <- indices[ord]

      n <- length(y_sorted)
      if (n < 2 * min_samples_leaf) return(NULL)

      # candidate split positions where x changes (between k and k+1)
      diff_mask <- x_sorted[-n] != x_sorted[-1]
      if (!any(diff_mask)) return(NULL)
      split_pos <- which(diff_mask)  # k in 1..n-1

      # leaf size constraints
      left_n <- split_pos
      right_n <- n - split_pos
      valid <- (left_n >= min_samples_leaf) & (right_n >= min_samples_leaf)
      split_pos <- split_pos[valid]
      if (length(split_pos) == 0) return(NULL)

      # cumulative sums for fast SSE
      csum <- cumsum(y_sorted)
      csum2 <- cumsum(y_sorted * y_sorted)
      total_sum <- csum[n]
      total_sum2 <- csum2[n]

      left_sum <- csum[split_pos]
      left_sum2 <- csum2[split_pos]
      right_sum <- total_sum - left_sum
      right_sum2 <- total_sum2 - left_sum2

      left_n_f <- as.numeric(split_pos)
      right_n_f <- as.numeric(n - split_pos)

      left_sse <- left_sum2 - (left_sum * left_sum) / left_n_f
      right_sse <- right_sum2 - (right_sum * right_sum) / right_n_f

      within_sse <- left_sse + right_sse
      gains <- parent_sse - within_sse

      best_i <- which.max(gains)
      best_gain <- gains[best_i]
      if (is.na(best_gain) || best_gain <= 0) return(NULL)

      k <- split_pos[best_i]
      split_value <- (x_sorted[k] + x_sorted[k + 1]) / 2.0

      denom <- within_sse[best_i] / max(n - 2, 1)
      f_stat <- if (denom > 0) best_gain / denom else Inf

      left_idx <- idx_sorted_global[1:k]
      right_idx <- idx_sorted_global[(k + 1):n]

      list(
        split_value = as.numeric(split_value),
        gain = as.numeric(best_gain),
        f_stat = as.numeric(f_stat),
        left_idx = left_idx,
        right_idx = right_idx
      )
    },

    predict = function(X) {
      if (is.null(root_)) stop("Model not fitted")
      if (is.data.frame(X)) X <- as.matrix(X)
      X <- as.matrix(X)
      storage.mode(X) <- "double"

      preds <- numeric(nrow(X))
      preds <- predict_recursive(root_, X, seq_len(nrow(X)), preds)
      preds
    },

    predict_recursive = function(node, X, row_idx, results) {
      if (length(row_idx) == 0) return(results)

      if (node$is_leaf) {
        results[row_idx] <- node$prediction
        return(results)
      }

      col <- node$split_feature_idx
      thr <- node$split_value

      xcol <- X[row_idx, col]
      goes_left <- xcol <= thr

      results <- predict_recursive(node$left, X, row_idx[goes_left], results)
      results <- predict_recursive(node$right, X, row_idx[!goes_left], results)
      results
    },

    mse = function(X, y) {
      y <- as.numeric(y)
      y_pred <- predict(X)
      mean((y - y_pred)^2)
    },

    score = function(X, y) {
      y <- as.numeric(y)
      y_pred <- predict(X)
      ss_res <- sum((y - y_pred)^2)
      ss_tot <- sum((y - mean(y))^2)
      if (ss_tot > 0) 1 - (ss_res / ss_tot) else 0.0
    },

    print_tree = function(max_depth_print = NA) {
      if (is.null(root_)) {
        cat("Model not fitted\n")
        return(invisible(NULL))
      }

      print_node <- function(node) {
        if (!is.na(max_depth_print) && node$depth > max_depth_print) return()

        indent <- paste(rep("  ", node$depth), collapse = "")
        if (node$is_leaf) {
          cat(sprintf("%sLeaf: pred=%.4f, sse=%.4f, n=%d\n", indent, node$prediction, node$sse, node$n_samples))
        } else {
          fname <- feature_names_[node$split_feature_idx]
          cat(sprintf("%s%s <= %.6f (gain=%.6f, F=%.2f, n=%d)\n",
                      indent, fname, node$split_value, node$split_gain, node$f_statistic, node$n_samples))
          print_node(node$left)
          print_node(node$right)
        }
      }

      print_node(root_)
      invisible(NULL)
    },

    summary = function() {
      cat("AID Regressor\n")
      cat(sprintf("min_samples_split=%d | min_samples_leaf=%d | max_depth=%d\n",
                  as.integer(min_samples_split), as.integer(min_samples_leaf), as.integer(max_depth)))
      cat(sprintf("min_split_gain=%.6g | max_features=%s | random_state=%d\n",
                  min_split_gain,
                  ifelse(max_features >= n_features_, "all", as.character(as.integer(max_features))),
                  as.integer(random_state)))
      cat(sprintf("Number of leaves: %d | Stored splits: %d\n", as.integer(n_leaves_), length(history_)))
      invisible(NULL)
    }
  )
)

File: AID/python/aid.py
#AID
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, List, Dict, Any, Union, Sequence, Tuple
import json
import numpy as np

ArrayLike = Union[np.ndarray, Sequence[Sequence[float]], Sequence[float]]


def _ensure_2d_float(X: ArrayLike) -> np.ndarray:
    X = np.asarray(X, dtype=float)
    if X.ndim == 1:
        X = X.reshape(-1, 1)
    if X.ndim != 2:
        raise ValueError("X must be 2D.")
    return X


def _ensure_1d_float(y: ArrayLike) -> np.ndarray:
    y = np.asarray(y, dtype=float)
    if y.ndim != 1:
        y = y.reshape(-1)
    return y


def _compute_sse(sum_y: float, sum_y2: float, n: int) -> float:
    """SSE = sum(y^2) - (sum(y)^2)/n"""
    if n <= 0:
        return 0.0
    return float(sum_y2 - (sum_y * sum_y) / n)


def _resolve_max_features(max_features, n_features: int) -> int:
    if max_features is None:
        return n_features
    if isinstance(max_features, str):
        key = max_features.lower()
        if key == "sqrt":
            return max(1, int(np.sqrt(n_features)))
        if key == "log2":
            return max(1, int(np.log2(n_features)))
        raise ValueError("max_features string must be one of: None, 'sqrt', 'log2'")
    if isinstance(max_features, float):
        if not (0.0 < max_features <= 1.0):
            raise ValueError("max_features float must be in (0, 1].")
        return max(1, int(np.ceil(max_features * n_features)))
    # int-like
    k = int(max_features)
    if k <= 0:
        raise ValueError("max_features must be >= 1")
    return min(k, n_features)


@dataclass
class _SplitCandidate:
    feature_idx: int
    split_value: float
    gain: float
    f_stat: float
    left_indices: np.ndarray   # global indices
    right_indices: np.ndarray  # global indices


class AIDNode:
    __slots__ = (
        "depth", "n_samples", "sse", "prediction", "sum_y", "sum_y2",
        "split_feature_idx", "split_value", "gain", "f_stat",
        "left", "right"
    )

    def __init__(self, depth: int, n_samples: int, sse: float, prediction: float,
                 sum_y: float, sum_y2: float):
        self.depth = depth
        self.n_samples = n_samples
        self.sse = sse
        self.prediction = prediction
        self.sum_y = sum_y
        self.sum_y2 = sum_y2

        self.split_feature_idx: Optional[int] = None
        self.split_value: Optional[float] = None
        self.gain: float = 0.0
        self.f_stat: Optional[float] = None
        self.left: Optional["AIDNode"] = None
        self.right: Optional["AIDNode"] = None

    @property
    def is_leaf(self) -> bool:
        return self.split_feature_idx is None


class AID:
    """
    AID regressor using SSE reduction (least squares) splits.

    Parameters
    ----------
    min_samples_leaf : int, default=5
        Minimum samples per leaf (AID parameter R).
    min_samples_split : int, default=10
        Minimum samples in a node to attempt a split (AID parameter M).
    max_depth : int, default=5
        Maximum depth (root depth = 0) (AID parameter Q).
    min_gain : float, default=0.0
        Minimum SSE reduction required to accept a split.
    max_leaves : int, default=10**9
        Optional cap on number of leaves.
    presort : bool, default=True
        If True, presort each feature once at fit and filter via node mask.
    store_history : bool, default=False
        Store split history.
    max_features : None | int | float | {"sqrt","log2"}, default=None
        Feature subsampling per node (like typical trees). If None uses all features.
    random_state : int | None, default=0
        RNG seed for feature subsampling (only relevant if max_features < n_features).
    """

    def __init__(
        self,
        min_samples_leaf: int = 5,
        min_samples_split: int = 10,
        max_depth: int = 5,
        min_gain: float = 0.0,
        store_history: bool = False,
        max_leaves: int = 10**9,
        presort: bool = True,
        max_features=None,
        random_state: Optional[int] = 0,
    ):
        self.min_samples_leaf = int(min_samples_leaf)
        self.min_samples_split = int(min_samples_split)
        self.max_depth = int(max_depth)
        self.min_gain = float(min_gain)
        self.store_history = bool(store_history)
        self.max_leaves = int(max_leaves)
        self.presort = bool(presort)
        self.max_features = max_features
        self.random_state = random_state

        self.root_: Optional[AIDNode] = None
        self.history_: List[Dict[str, Any]] = []
        self._n_leaves = 0

        # fitted state
        self._X: Optional[np.ndarray] = None
        self._y: Optional[np.ndarray] = None
        self._sorted_indices: Optional[List[np.ndarray]] = None  # per feature sorted sample indices
        self.feature_names_: Optional[List[str]] = None
        self.n_features_: int = 0

        self._rng = np.random.default_rng(random_state)

    # -----------------------------
    # Fit
    # -----------------------------
    def fit(self, X: ArrayLike, y: ArrayLike, feature_names: Optional[List[str]] = None) -> "AID":
        X_arr, y_arr = self._validate_input(X, y)

        self._X = X_arr
        self._y = y_arr
        self.n_features_ = X_arr.shape[1]
        self.feature_names_ = feature_names if feature_names is not None else [f"X{i}" for i in range(X_arr.shape[1])]

        if self.presort:
            self._presort_numeric_features()
        else:
            self._sorted_indices = None

        indices = np.arange(y_arr.size, dtype=int)
        sum_y = float(np.sum(y_arr))
        sum_y2 = float(np.sum(y_arr * y_arr))
        sse = _compute_sse(sum_y, sum_y2, int(y_arr.size))
        prediction = float(sum_y / max(int(y_arr.size), 1))

        self.root_ = AIDNode(depth=0, n_samples=int(y_arr.size), sse=sse,
                             prediction=prediction, sum_y=sum_y, sum_y2=sum_y2)

        self.history_.clear()
        self._n_leaves = 0

        self.root_ = self._build_tree(self.root_, indices, depth=0)
        return self

    def _validate_input(self, X, y) -> Tuple[np.ndarray, np.ndarray]:
        X_arr = _ensure_2d_float(X)
        y_arr = _ensure_1d_float(y)

        if X_arr.shape[0] != y_arr.shape[0]:
            raise ValueError(f"Shape mismatch: X has {X_arr.shape[0]} samples, y has {y_arr.shape[0]}")

        if np.any(np.isnan(X_arr)) or np.any(np.isinf(X_arr)):
            raise ValueError("X contains NaN or Inf")

        if np.any(np.isnan(y_arr)) or np.any(np.isinf(y_arr)):
            raise ValueError("y contains NaN or Inf")

        return X_arr, y_arr

    def _presort_numeric_features(self):
        assert self._X is not None
        self._sorted_indices = [
            np.argsort(self._X[:, j], kind="mergesort") for j in range(self.n_features_)
        ]

    def _should_stop(self, node: AIDNode, depth: int) -> bool:
        if depth >= self.max_depth:
            return True
        if node.n_samples < self.min_samples_split:
            return True
        if self._n_leaves >= self.max_leaves:
            return True
        if node.n_samples < 2 * self.min_samples_leaf:
            return True
        return False

    # -----------------------------
    # Split search (indices-based)
    # -----------------------------
    def _feature_subset(self) -> np.ndarray:
        k = _resolve_max_features(self.max_features, self.n_features_)
        if k >= self.n_features_:
            return np.arange(self.n_features_, dtype=int)
        # deterministic given RNG state
        return self._rng.choice(self.n_features_, size=k, replace=False)

    def _find_best_split(self, indices: np.ndarray, parent_node: AIDNode) -> Optional[_SplitCandidate]:
        assert self._X is not None and self._y is not None
        X = self._X
        y = self._y

        n = indices.size
        if n < 2 * self.min_samples_leaf:
            return None

        # mask for fast filtering (only if presort enabled)
        if self._sorted_indices is not None:
            in_node = np.zeros(y.size, dtype=bool)
            in_node[indices] = True

        best_candidate: Optional[_SplitCandidate] = None
        features = self._feature_subset()

        for feature_idx in features:
            if self._sorted_indices is not None:
                sorted_full = self._sorted_indices[feature_idx]
                relevant = sorted_full[in_node[sorted_full]]
            else:
                relevant = indices[np.argsort(X[indices, feature_idx], kind="mergesort")]

            if relevant.size < 2 * self.min_samples_leaf:
                continue

            x_sorted = X[relevant, feature_idx]
            y_sorted = y[relevant]

            # candidate split positions where x changes
            diff_mask = x_sorted[:-1] != x_sorted[1:]
            if not np.any(diff_mask):
                continue

            split_positions = np.where(diff_mask)[0]  # split between pos and pos+1

            # enforce min_samples_leaf
            left_n = split_positions + 1
            right_n = relevant.size - left_n
            valid_mask = (left_n >= self.min_samples_leaf) & (right_n >= self.min_samples_leaf)
            split_positions = split_positions[valid_mask]
            if split_positions.size == 0:
                continue

            csum_y = np.cumsum(y_sorted)
            csum_y2 = np.cumsum(y_sorted * y_sorted)
            total_sum = csum_y[-1]
            total_sum2 = csum_y2[-1]

            left_sum = csum_y[split_positions]
            left_sum2 = csum_y2[split_positions]
            right_sum = total_sum - left_sum
            right_sum2 = total_sum2 - left_sum2

            left_n_float = (split_positions + 1).astype(float)
            right_n_float = (relevant.size - (split_positions + 1)).astype(float)

            left_sse = left_sum2 - (left_sum * left_sum) / left_n_float
            right_sse = right_sum2 - (right_sum * right_sum) / right_n_float

            within_sse = left_sse + right_sse
            gains = parent_node.sse - within_sse

            best_idx = int(np.argmax(gains))
            best_gain = float(gains[best_idx])
            if best_gain <= 0:
                continue

            # "F-stat" descriptive (gain divided by within/(n-2))
            denom = float(within_sse[best_idx] / max(relevant.size - 2, 1))
            f_stat = float(best_gain / denom) if denom > 0 else float("inf")

            split_pos = int(split_positions[best_idx])
            split_value = float((x_sorted[split_pos] + x_sorted[split_pos + 1]) / 2.0)

            left_indices = relevant[: split_pos + 1]
            right_indices = relevant[split_pos + 1 :]

            candidate = _SplitCandidate(
                feature_idx=int(feature_idx),
                split_value=split_value,
                gain=best_gain,
                f_stat=f_stat,
                left_indices=left_indices,
                right_indices=right_indices,
            )

            if best_candidate is None or candidate.gain > best_candidate.gain:
                best_candidate = candidate

        return best_candidate

    # -----------------------------
    # Tree growth (indices-based)
    # -----------------------------
    def _build_tree(self, node: AIDNode, indices: np.ndarray, depth: int) -> AIDNode:
        if self._should_stop(node, depth):
            self._n_leaves += 1
            return node

        candidate = self._find_best_split(indices, node)
        if candidate is None or candidate.gain <= self.min_gain:
            self._n_leaves += 1
            return node

        assert self._y is not None
        y = self._y

        left_indices = candidate.left_indices
        right_indices = candidate.right_indices

        if left_indices.size < self.min_samples_leaf or right_indices.size < self.min_samples_leaf:
            self._n_leaves += 1
            return node

        left_y = y[left_indices]
        right_y = y[right_indices]

        left_sum = float(np.sum(left_y))
        left_sum2 = float(np.sum(left_y * left_y))
        right_sum = float(np.sum(right_y))
        right_sum2 = float(np.sum(right_y * right_y))

        left_sse = _compute_sse(left_sum, left_sum2, int(left_y.size))
        right_sse = _compute_sse(right_sum, right_sum2, int(right_y.size))

        node.split_feature_idx = candidate.feature_idx
        node.split_value = candidate.split_value
        node.gain = candidate.gain
        node.f_stat = candidate.f_stat

        if self.store_history:
            self.history_.append(
                dict(
                    depth=depth,
                    feature_idx=candidate.feature_idx,
                    feature_name=self.feature_names_[candidate.feature_idx] if self.feature_names_ else f"X{candidate.feature_idx}",
                    split_value=candidate.split_value,
                    gain=candidate.gain,
                    f_stat=candidate.f_stat,
                    parent_n=node.n_samples,
                    parent_mean=node.prediction,
                    parent_sse=node.sse,
                    n_left=int(left_y.size),
                    mean_left=float(left_sum / left_y.size),
                    sse_left=float(left_sse),
                    n_right=int(right_y.size),
                    mean_right=float(right_sum / right_y.size),
                    sse_right=float(right_sse),
                )
            )

        node.left = AIDNode(
            depth=depth + 1,
            n_samples=int(left_y.size),
            sse=float(left_sse),
            prediction=float(left_sum / left_y.size),
            sum_y=left_sum,
            sum_y2=left_sum2,
        )
        node.right = AIDNode(
            depth=depth + 1,
            n_samples=int(right_y.size),
            sse=float(right_sse),
            prediction=float(right_sum / right_y.size),
            sum_y=right_sum,
            sum_y2=right_sum2,
        )

        node.left = self._build_tree(node.left, left_indices, depth + 1)
        node.right = self._build_tree(node.right, right_indices, depth + 1)
        return node

    # -----------------------------
    # Predict
    # -----------------------------
    def predict(self, X: ArrayLike) -> np.ndarray:
        if self.root_ is None:
            raise ValueError("Model not fitted")
        X_arr = _ensure_2d_float(X)

        predictions = np.empty(X_arr.shape[0], dtype=float)
        stack: List[Tuple[AIDNode, np.ndarray]] = [(self.root_, np.arange(X_arr.shape[0], dtype=int))]

        while stack:
            node, rows = stack.pop()
            if rows.size == 0:
                continue

            if node.is_leaf:
                predictions[rows] = node.prediction
                continue

            feature_idx = int(node.split_feature_idx)  # type: ignore[arg-type]
            split_value = float(node.split_value)      # type: ignore[arg-type]
            x_col = X_arr[rows, feature_idx]
            goes_left = x_col <= split_value

            left_rows = rows[goes_left]
            right_rows = rows[~goes_left]

            stack.append((node.left, left_rows))   # type: ignore[arg-type]
            stack.append((node.right, right_rows)) # type: ignore[arg-type]

        return predictions

    # -----------------------------
    # Score / utilities
    # -----------------------------
    def score(self, X: ArrayLike, y: ArrayLike) -> float:
        y_arr = _ensure_1d_float(y)
        y_pred = self.predict(X)

        ss_res = float(np.sum((y_arr - y_pred) ** 2))
        ss_tot = float(np.sum((y_arr - np.mean(y_arr)) ** 2))
        return float(1 - (ss_res / ss_tot)) if ss_tot > 0 else 0.0

    def mse(self, X: ArrayLike, y: ArrayLike) -> float:
        y_arr = _ensure_1d_float(y)
        y_pred = self.predict(X)
        return float(np.mean((y_arr - y_pred) ** 2))

    def summary(self) -> str:
        if self.root_ is None:
            return "AID(not fitted)"
        n_splits = len(self.history_) if self.store_history else "N/A"
        return (
            f"AID(min_samples_leaf={self.min_samples_leaf}, "
            f"min_samples_split={self.min_samples_split}, "
            f"max_depth={self.max_depth}, min_gain={self.min_gain}, presort={self.presort}, "
            f"max_features={self.max_features}, random_state={self.random_state})\n"
            f"Root: n={self.root_.n_samples}, mean={self.root_.prediction:.6f}, sse={self.root_.sse:.6f}\n"
            f"Splits stored: {n_splits}"
        )

    def print_tree(self, max_depth: Optional[int] = None):
        if self.root_ is None:
            print("Model not fitted")
            return

        def print_node(node: AIDNode, depth: int = 0):
            if max_depth is not None and depth > max_depth:
                return

            indent = "  " * depth

            if node.is_leaf:
                print(f"{indent}Leaf: mean={node.prediction:.3f}, sse={node.sse:.3f}, n={node.n_samples}")
            else:
                feature_name = self.feature_names_[node.split_feature_idx] if self.feature_names_ else f"X{node.split_feature_idx}"
                print(f"{indent}{feature_name} <= {node.split_value:.6f} (gain={node.gain:.6f}, F={node.f_stat:.2f}, n={node.n_samples})")
                print_node(node.left, depth + 1)   # type: ignore[arg-type]
                print_node(node.right, depth + 1)  # type: ignore[arg-type]

        print_node(self.root_, 0)

    def to_dict(self) -> Dict[str, Any]:
        if self.root_ is None:
            raise ValueError("Model not fitted")

        def node_to_dict(node: AIDNode) -> Dict[str, Any]:
            d = {
                "depth": node.depth,
                "n_samples": node.n_samples,
                "prediction": node.prediction,
                "sse": node.sse,
            }
            if node.is_leaf:
                d["is_leaf"] = True
            else:
                d["is_leaf"] = False
                d["split_feature_idx"] = node.split_feature_idx
                d["split_feature_name"] = self.feature_names_[node.split_feature_idx] if self.feature_names_ else f"X{node.split_feature_idx}"
                d["split_value"] = node.split_value
                d["gain"] = node.gain
                d["f_stat"] = node.f_stat
                d["left"] = node_to_dict(node.left)   # type: ignore[arg-type]
                d["right"] = node_to_dict(node.right) # type: ignore[arg-type]
            return d

        return {
            "params": {
                "min_samples_leaf": self.min_samples_leaf,
                "min_samples_split": self.min_samples_split,
                "max_depth": self.max_depth,
                "min_gain": self.min_gain,
                "presort": self.presort,
                "max_features": self.max_features,
                "random_state": self.random_state,
            },
            "tree": node_to_dict(self.root_),
            "history": self.history_ if self.store_history else None,
        }

    def to_json(self, path: str):
        obj = self.to_dict()
        with open(path, "w", encoding="utf-8") as f:
            json.dump(obj, f, indent=2)
